text/FilipBartodziej/opis.tex

```diff
@@ -0,0 +1,188 @@
+\EndFor
+\State update groups in $Ssubstr$ using old $Ssubstr$ and $SsubstrPrefixes[d]$ 
+\EndFor
+\State $s' \gets$ map $Ssubstr$ to its buckets number, ordered as in $s$
+\State \Return $unmap(small-large(s'))$
+\EndProcedure
+\end{algorithmic}
+
+Po pętli for S podsłowa są uporządkowane w kolejności prawie leksyko-graficznej. Jedyny wyjątek to gdy $\alpha$ jest podsłowem $\beta$, wtedy $\alpha > \beta$ w tym porządku.
+
+Oznaczmy jako $T'_i$ sufiks w $s'$ odpowiadający S sufiksowi $T_i$. Poprawność sortowania wynika z poniższego faktu.
+
+\begin{lemma}
+$T_i < T_j \iff T'_i < T'_j$
+\end{lemma}
+\begin{proof}
+\item[$T_i < T_j \Rightarrow T'_i < T'_j$] - trywialnie z mapowania
+\item[$T_i < T_j \Leftarrow T'_i < T'_j$] - opis idei\\
+Rozważa się pierwszy znak różniący $T'_i < T'_j$. Jest to numer jakiejś grupy, która indukuje S podsłowo. Odzyskujemy te podsłowa $\alpha,\beta$ i rozważamy dwa przypadki - $\beta$ to prefiks $\alpha$ lub nie. Z tego można wywnioskować $T_i < T_j$.
+\end{proof}
+
+
+\section*{Sortowanie całości - faza 2}
+\begin{algorithmic}
+\Procedure{finishS}{s, SL, sortedS}
+\State $A \gets$ bucket and sort $s$ suffixes by first character
+\For{$x$ in $reversed(sortedS)$}\Comment{\cref{SL_order}}
+\State move $x$ to the end of its bucket in $A$
+\State move that buckets end to the left by one
+\EndFor
+\For{$i \gets 1,...,n$ }\Comment{\cref{phase2-invariant}}
+\If{$SL[A[i]-1] == "L"$}
+\State move $A[i]-1$ to the front of its bucket
+\State move that buckets front to the right by one
+\EndIf
+\EndFor
+\State \Return A
+\EndProcedure
+\end{algorithmic}
+
+Po pierwszej pętli sufiksy typu S są już na poprawnych pozycjach. Pętla druga porządkuje resztę.
+
+\begin{lemma}\label{phase2-invariant}
+W kroku i drugiej pętli algorytmu, sufiks $T_{A[i]}$ jest już na poprawnej pozycji w tablicy sufiksów.
+\end{lemma}
+\begin{proof}
+Indukcja.
+\item[Baza indukcji:] wynika z tego, że najmniejszy sufiks jest typu S wiec jest już na poprawnej pozycji
+\item[Krok indukcyjny $i \rightarrow i+1$:] nie wprost - istnieje sufiks $k > i+1$, który w tablicy sufiksów powinien zająć miejsce $A[i+1]$.\\
+Wprost z założenia $T_{A[k]} < T_{A[i+1]}$ i obydwa te sufiksy są typu L - sufiksy S są na poprawnych miejscach.\\
+$T_{A[k]}$ i $T_{A[i+1]}$ muszą rozpoczynać się tą samą literą c - przez wcześniejsze dzielenie na grupy przez pierwszą literę $T_{A[i+1]}$ ma jako pierwszą najmniejszą literę z sufiksów $A[i+1],A[i+2],...,A[n]$. Niech $T_{A[i+1]}=c \alpha$ i $T_{A[k]}=c \beta$.\\
+\begin{gather*}
+T_{A[k]}\ is\ type\ L \Rightarrow \beta < T_{A[k]}\\
+T_{A[k]} < T_{A[i+1]} \Rightarrow \beta < \alpha\\
+(T_{A[k]}\ is\ A[i+1]th\ in\ suffix\ array) \land \beta < T_{A[k]} \Rightarrow \beta \in \{A[1],A[2],...,A[i]\}
+\end{gather*}
+$\beta < \alpha$, więc $T_{A[k]}$ był przesunięty na początek grupy (w branchy if) przed $T_{A[i+1]}$. Sufiksy te należą do tej samej grupy, zatem $T_{A[k]}$ leży na lewo od $T_{A[i+1]}$, sprzeczność.
```

Comment:
"w gałęzi if" - even if we allow English in pseudocodes, then still it'd better to use Polish in all other cases ;)